File: 02-neural-machine-translation-transformer/src/language_detection.py
from __future__ import annotations

import re
from dataclasses import dataclass

from .text_preprocessing import clean_text

DEVANAGARI_RE = re.compile(r"[\u0900-\u097F]")
LATIN_RE = re.compile(r"[A-Za-z]")


@dataclass(frozen=True)
class LanguageDetection:
    language: str
    latin_characters: int
    devanagari_characters: int
    latin_ratio: float
    devanagari_ratio: float
    reason: str

    def as_dict(self) -> dict[str, object]:
        return {
            "language": self.language,
            "latin_characters": self.latin_characters,
            "devanagari_characters": self.devanagari_characters,
            "latin_ratio": round(self.latin_ratio, 4),
            "devanagari_ratio": round(self.devanagari_ratio, 4),
            "reason": self.reason,
        }


def detect_language(
    text: str,
    *,
    mixed_script_min_ratio: float = 0.20,
    minimum_letters: int = 2,
) -> LanguageDetection:
    cleaned = clean_text(text)
    latin = len(LATIN_RE.findall(cleaned))
    devanagari = len(DEVANAGARI_RE.findall(cleaned))
    total = latin + devanagari

    if total < minimum_letters:
        return LanguageDetection(
            "uncertain",
            latin,
            devanagari,
            0.0 if total == 0 else latin / total,
            0.0 if total == 0 else devanagari / total,
            "Too little Latin or Devanagari evidence.",
        )

    latin_ratio = latin / total
    devanagari_ratio = devanagari / total

    if latin and devanagari:
        if min(latin_ratio, devanagari_ratio) >= mixed_script_min_ratio:
            return LanguageDetection(
                "mixed",
                latin,
                devanagari,
                latin_ratio,
                devanagari_ratio,
                "Meaningful Latin and Devanagari content is present.",
            )
        dominant = "english" if latin_ratio > devanagari_ratio else "hindi"
        return LanguageDetection(
            dominant,
            latin,
            devanagari,
            latin_ratio,
            devanagari_ratio,
            f"{dominant.title()} script is strongly dominant.",
        )

    language = "english" if latin else "hindi"
    return LanguageDetection(
        language,
        latin,
        devanagari,
        latin_ratio,
        devanagari_ratio,
        f"Only {'Latin' if language == 'english' else 'Devanagari'} letters detected.",
    )

File: 02-neural-machine-translation-transformer/web/src/language-detection.js
const DEVANAGARI = /[\u0900-\u097F]/u;
const LATIN = /[A-Za-z]/u;

export function detectLanguage(text) {
  const value = String(text ?? '').normalize('NFKC').trim();
  let devanagari = 0;
  let latin = 0;

  for (const char of value) {
    if (DEVANAGARI.test(char)) devanagari += 1;
    else if (LATIN.test(char)) latin += 1;
  }

  const letters = devanagari + latin;
  if (letters < 2) {
    return { language: 'uncertain', devanagari, latin, letters, confidence: 0 };
  }

  const devanagariRatio = devanagari / letters;
  const latinRatio = latin / letters;
  if (devanagari >= 2 && latin >= 2 && devanagariRatio >= 0.2 && latinRatio >= 0.2) {
    return { language: 'mixed', devanagari, latin, letters, confidence: Math.max(devanagariRatio, latinRatio) };
  }
  if (devanagariRatio >= 0.65) {
    return { language: 'hindi', devanagari, latin, letters, confidence: devanagariRatio };
  }
  if (latinRatio >= 0.65) {
    return { language: 'english', devanagari, latin, letters, confidence: latinRatio };
  }
  return { language: 'uncertain', devanagari, latin, letters, confidence: Math.max(devanagariRatio, latinRatio) };
}

export function resolveDirection(text, selected = 'auto') {
  const detection = detectLanguage(text);
  if (selected === 'en-hi' || selected === 'hi-en') {
    return { direction: selected, detection };
  }
  if (detection.language === 'english') return { direction: 'en-hi', detection };
  if (detection.language === 'hindi') return { direction: 'hi-en', detection };
  return { direction: null, detection };
}

export function directionLabel(direction) {
  return direction === 'en-hi' ? 'English → Hindi' : direction === 'hi-en' ? 'Hindi → English' : '—';
}
